docs/contributing/coding-style-guide.md

```diff
@@ -0,0 +1,665 @@
+        messages = messages[:0]
+    }
+}
+```
+
+</td></tr>
+</tbody></table>
+
+### Readability
+
+#### Only Two Ways of Formatting Functions/Methods
+
+Prefer function/method definitions with arguments in a single line. If it's too wide, put each argument on a new line.
+
+<table>
+<thead align="center"><tr><th>Avoid 🔥</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+func function(argument1 int, argument2 string,
+    argument3 time.Duration, argument4 someType,
+    argument5 float64, argument6 time.Time,
+) (ret int, err error) {
+```
+
+</td></tr>
+</tbody></table>
+<table>
+<thead align="center"><tr><th>Better 🤓</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+func function(
+    argument1 int,
+    argument2 string,
+    argument3 time.Duration,
+    argument4 someType,
+    argument5 float64,
+    argument6 time.Time,
+) (ret int, err error)
+```
+
+</td></tr>
+</tbody></table>
+
+#### Control Structure: Avoid `else`
+
+In most of the cases you don't need `else`. You can usually use `continue`, `break` or `return` to end an `if` block.
+This enables having one less indent and netter consistency so code is more readable.
+
+<table>
+<thead align="center"><tr><th>Avoid 🔥</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+for _, elem := range elems {
+    if a == 1 {
+        something[i] = "yes"
+    } else
+        something[i] = "no"
+    }
+}
+```
+
+</td></tr>
+</tbody></table>
+<table>
+<thead align="center"><tr><th>Better 🤓</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+for _, elem := range elems {
+    if a == 1 {
+        something[i] = "yes"
+        continue
+    }
+    something[i] = "no"
+}
+```
+
+</td></tr>
+</tbody></table>
+
+#### Use Named Return Parameters Carefully
+
+It's OK to name return parameters if the types do not give enough information about what function or method actually returns.
+Another use case is when you want to define a variable, e.g. a slice.
+
+**IMPORTANT:** never use naked `return` statements with named return perameters. This compiles but it makes returning values
+implicit and thus more prone to surprises.
+
+#### Explicitly Handled Return Errors
+
+Always address returned errors. It does not mean you cannot "ignore" the error for some reason, e.g. if we know implementation
+will not return anything meaningful. You can ignore the error, but do so explicitly:
+
+<table>
+<thead align="center"><tr><th>Avoid 🔥</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+someMethodThatReturnsError(...)
+```
+
+</td></tr>
+</tbody></table>
+<table>
+<thead align="center"><tr><th>Better 🤓</th></tr></thead>
+<tbody>
+<tr><td>
+
+
+```go
+_ = someMethodThatReturnsError(...)
+```
+
+</td></tr>
+</tbody></table>
+
+The exception: well known cases such as `level.Debug|Warn` etc and `fmt.FPrint*`
+
+#### Wrap Errors for More Context; Don't Repeat "failed ..." There.
+
+We use [`pkg/errors`](https://github.com/pkg/errors) package for `errors`. We prefer it over standard wrapping with `fmt.Sprintf` + `%w`,
```

Comment:
I would disagree here. I don't think it's that bad and there are linters to check for %w usage (something by Fatih). Additionally, I would prefer using upstream. IIRC errors package itself is starting to use the new errors API on newer Go version under the hood.

Reply:
AFAIK `pkg/errors` started to support new errors API, not using it under the hood. Also `pkg/errors` still offers more feature, like error stacks which Thanos benefits a lot, I believe. The std errors not there yet. However, there's nothing wrong to combine them, just preference.

Reply:
Good point on more features. I guess for a project as invested into `pkg/errors` and still being compatible with stdlib it makes sense to keep using it.
For new projects, I started using only stdlib errors. :)

Reply:
Thanks for starting this discussion, I remember you like the standard one. (: There are lots of controversial things around the standard `errors` package. I am personally not a fan of the new changes to the `errors` and the same is with the community (I remember @peterbourgon was tweeting about it). I am happy to elaborate more on why the standard package is to me, just wrong and worse (for now) than `pkg/errors`:

1. You are maybe right that the are some linters for `%w`. But I would argue that something assured on compile time is always better than run time or pre run-time static analysis.
1. As I mentioned you can easily end up with inconsistent wraps. A developer can put `%w` literally **everywhere**. You can do:
   * `fmt.Errorf("failed to close file: %w, err)`
   * `fmt.Errorf("failed to close file %w, err)`
   * `fmt.Errorf("failed to close file. Err: %w, err)`
   * ... thousands of other non lintable inconsistencies.
1. `pkg/errors` does not add many features, and TBH that's good to not have feature creep. The only features we want is really `%+v` which gives you stack trace for the error. 
1. I just find `fmt.*` too close to `fmt.Print` `fmt.Sprintf` for string formatting, not error wrapping. It just surprises me every time I read `fmt.Errors` : but it might be personal brain malfunction ;p
1. To me `errors` have two ways of doing the same thing: `As` and `Is`.  Having a single package with only a few functions and already such clear code smell... Is interesting. (: I might not see `As` use case, as well. I would love it if someone could educate me why such an odd decision was made, but probably there were some important facts I am missing was there.
1. Since it's standard and has to be compatible with Go 1.0 velocity of  `errors` is poor, so nothing important will be improved there. So to me being a standard package is not really a benefit ): 

On the contrary:

* `pkg/error` has similar flaws in terms of two ways for same thing e.g `Unwrap`)
* `pkg/errors` is still more popular, so any dependency which uses that will not work with `As` and `Is`... HOWEVER recently with the new version they reverted the most important change: https://github.com/pkg/errors/pull/220 :scream: (it conflicts with unused by as `Unwrap` method) This means that either:
* We should never update `pkg/errors` ;p (bad)
* We should just copy those few lines and use internally (bad)
* We should use well maintained https://github.com/friendsofgo/errors fork. (best)

So it's not black and white but TBH I would propose https://github.com/friendsofgo/errors  (: it looks like we trust those maintainers, they seem a bit more reasonable. But again, it's maybe bit bigger decision, so I would stick to `pkg/errors` for now as without upgrading to 0.9.0 it's really the same for us: https://github.com/friendsofgo/errors

Reply:
Like it or not, stdlib errors and wrapping with %w is the future. You should be using it.

Reply:
Hah. "Like or not, we will die at some point" (: 

Why not making future brighter? At least starting by our own code and explaining why the standard way is bit worse? (:

Reply:
> Gofmt's style is no one's favorite, yet gofmt is everyone's favorite.

I guess that might be true for the errors implementation going forward. Personally, I have no problems with it.
The discussion point was mostly a compatibility one. As Thanos is heavily invested in pkg/errors and there is interoperability with the new errors built-in functionality, I'm fine with sticking to pkg/errors for the meantime and evaluate in the future once more.

Reply:
> Why not making future brighter? At least starting by our own code and explaining why the standard way is bit worse?

I guess it's fine if you want to suggest a helper package that extends stdlib package errors with other stuff you want to use, as long as it doesn't wrap, unwrap, annotate, check, compare, etc. errors in an incompatible way. But you're still going to need to know how "normal" error wrapping, unwrapping, etc. works.

> To me errors have two ways of doing the same thing: As and Is.

Not exactly. [See here for more detail](http://peter.bourgon.org/blog/2019/09/11/programming-with-errors.html).

Reply:
Cool, can agree with you. Let's move this discussion to separate place, but I would vote for `errors.Wrap` extension that will be fully on top of std errors :hugs: